### PR TITLE
Filter inputs in input selection

### LINKED ISSUE #1198: [Task]: Handle nft outputs with expiration or storage deposit return unlock conditions
### Task description

Currently we only request nft outputs where the address unlock condition has an address from the account, but if it's only in the expiration or storage deposit return unlock condition, then it's not requested. So the additional request needs to be added and then we also need to check for outputs that can be controlled by this address, if we can unlock the nft output before adding them to the balance/available inputs.

For outputs with an alias address we might need to add the same checks (if nft/alias output isn't there, we have to mark them as not available (should currently also happening when requesting <https://github.com/iotaledger/wallet.rs/blob/d7597f6d7483dac6016c69834aa5ac563ec2bb8a/src/account/operations/syncing/addresses.rs#L125>, but might be good to also add in the balance calculation)

### Requirements

-

### Acceptance criteria

-

### Creation checklist

- [ ] I have assigned this task to the correct people
- [X] I have added the most appropriate labels
- [X] I have linked the correct milestone and/or project

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Filter outputs based on unlock conditions before adding them to the inputs from which we actually try to select the final inputs

## Links to any relevant issues

Fixes #1198 and #1118

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/account/operations/transaction/input_selection.rs [hunk 1]
```diff
@@ -63,65 +68,19 @@ impl AccountHandle {
...
-        }
+        // Filter inputs to not include inputs that require additional outputs for storage deposit return or could be
+        // still locked
+        let current_time = self.client.get_time_checked().await?;
+        let available_outputs_signing_data = filter_inputs(
+            &account.addresses_with_unspent_outputs,
+            account.unspent_outputs.values().collect(),
+            current_time,
+            &outputs,
+            account.locked_outputs.clone(),
```
Comment (thibault-martinez) [inline src/account/operations/transaction/input_selection.rs]: Do we really need ownership of this ?

File: src/account/operations/transaction/input_selection.rs [hunk 2]
```diff
@@ -63,65 +68,19 @@ impl AccountHandle {
...
-                    _ => {}
-                }
-            }
-        }
+        // Filter inputs to not include inputs that require additional outputs for storage deposit return or could be
+        // still locked
+        let current_time = self.client.get_time_checked().await?;
+        let available_outputs_signing_data = filter_inputs(
+            &account.addresses_with_unspent_outputs,
+            account.unspent_outputs.values().collect(),
```
Comment (thibault-martinez) [inline src/account/operations/transaction/input_selection.rs]: Do we really need ownership ?

File: src/account/operations/transaction/input_selection.rs [hunk 3]
```diff
@@ -150,3 +109,107 @@ impl AccountHandle {
...
+
+/// Filter available outputs to only include outputs that don't have unlock conditions, that could create
+/// conflicting transactions or need a new output for the storage deposit return
+/// Also only include Alias, Nft and Foundry outputs, if a corresponding output with the same id exists in the output,
+/// so they don't get burned
+///
+/// Note: this is only for the default input selection, it's still possible to send these outputs by using
+/// `claim_outputs` or providing their OutputId's in the custom_inputs
+///
+/// Some examples for which outputs should be inluded in the inputs to select from:
```
Comment (thibault-martinez) [inline src/account/operations/transaction/input_selection.rs]:
```suggestion
/// Some examples for which outputs should be included in the inputs to select from:
```

File: src/account/operations/transaction/input_selection.rs [hunk 4]
```diff
@@ -150,3 +109,107 @@ impl AccountHandle {
...
+fn filter_inputs(
+    addresses_with_unspent_outputs: &[AddressWithUnspentOutputs],
+    available_outputs: Vec<&OutputData>,
+    current_time: u32,
+    outputs: &[Output],
+    locked_outputs: HashSet<OutputId>,
+) -> crate::Result<Vec<InputSigningData>> {
+    let mut available_outputs_signing_data = Vec::new();
+    for output_data in available_outputs {
+        if !locked_outputs.contains(&output_data.output_id) {}
```
Comment (thibault-martinez) [inline src/account/operations/transaction/input_selection.rs]: ?

File: src/account/operations/transaction/input_selection.rs [hunk 5]
```diff
@@ -150,3 +109,110 @@ impl AccountHandle {
...
+
+        let output_can_be_unlocked_now_and_in_future = can_output_be_unlocked_forever_from_now_on(
+            // We use the addresses with unspent outputs, because other addresses of the
+            // account without unspent outputs can't be related to this output
+            addresses_with_unspent_outputs,
+            output_data,
+            current_time,
+        );
+
+        // Outputs that could get unlocked in the future will not be inluded
```
Comment (thibault-martinez) [inline src/account/operations/transaction/input_selection.rs]:
```suggestion
        // Outputs that could get unlocked in the future will not be included
```

**Review by samuel-rufi — DISMISSED** [review]

**Review by thibault-martinez — DISMISSED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Pull request merged**